glyphs-reader/src/font.rs

```diff
@@ -1630,20 +1648,65 @@ impl RawGlyph {
             }
             instances.push(layer.try_into()?);
         }
+        // if category/subcategory were set in the source, we keep them;
+        // otherwise we look them up based on the bundled GlyphData.
+        // (we use this info later to determine GDEF categories, zero the width
+        // on non-spacing marks, etc)
+        let mut category = self
+            .category
+            .map(|cat| Category::from_str(&cat))
+            .transpose()
+            .map_err(|category| Error::BadCategory {
+                glyph: self.glyphname.clone(),
+                category,
+            })?;
+        let mut sub_category = self
+            .sub_category
+            .map(|cat| Subcategory::from_str(&cat))
+            .transpose()
+            .map_err(|category| Error::BadCategory {
+                glyph: self.glyphname.clone(),
+                category,
+            })?;
+
+        let codepoints = self
+            .unicode
+            .map(|s| parse_codepoint_str(&s, codepoint_radix))
+            .unwrap_or_default();
+
+        if category.is_none() || sub_category.is_none() {
+            if let Some((computed_category, computed_subcategory)) =
+                get_glyph_category(&self.glyphname, &codepoints)
+            {
+                // if they were manually set don't change them, otherwise do
+                category = category.or(Some(computed_category));
+                sub_category = sub_category.or(Some(computed_subcategory));
+            }
+        }
+
         Ok(Glyph {
             glyphname: self.glyphname,
             export: self.export.unwrap_or(true),
             layers: instances,
             left_kern: self.kern_left,
             right_kern: self.kern_right,
-            unicode: self
-                .unicode
-                .map(|s| parse_codepoint_str(&s, codepoint_radix))
-                .unwrap_or_default(),
+            unicode: codepoints,
+            category,
+            sub_category,
         })
     }
 }
 
+// This will eventually need to be replaced with something that can handle
+// custom GlyphData.xml files, as well as handle overrides that are part of the
+// glyph source.
+fn get_glyph_category(name: &str, codepoints: &BTreeSet<u32>) -> Option<(Category, Subcategory)> {
```

Comment:
or maybe link to the py equivalent here?

Reply:
this is just a helper/wrapper around `get_glyph`, and doesn't have a python equivalent. `get_glyph` does, though, and that's documented below.